### PR TITLE
[SQL] SPARK-6548: Adding stddev to DataFrame functions

### PR DESCRIPTION
**Pull request by dreamquster**
remerge SPARK-6548
https://github.com/apache/spark/pull/5228

### CODE REVIEW

**Comment by dreamquster** [comment]
What's wrong?Who can verify it?

**Comment by marmbrus** [comment]
ok to test

**Comment by marmbrus** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -334,6 +334,24 @@ class DataFrameSuite extends QueryTest {
...
+      Row(testData2ADev))
+
+    checkAnswer(
+      testData2.agg(stddev(testData2("a"))),
+      Row(testData2ADev))
+
+    checkAnswer(
+      decimalData.agg(stddev('a), sumDistinct('a)), // non-partial
+      Row(testData2ADev, new java.math.BigDecimal(6)) :: Nil)
+
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: can you remove the excessive white space here?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -344,6 +344,20 @@ object functions {
    */
   def abs(e: Column): Column = Abs(e.expr)
 
+  /**
+   * Aggregate func:compute the stddev of the specific column.
+   *
+   * @group agg_funcs
+   */
+  def stddev(columnName: String): Column = stddev(Column(columnName))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: to be consistent with other functiosn, can we just use "e"?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: ah never mind -- was lookin at the wrong one

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 3]
```diff
@@ -344,6 +344,20 @@ object functions {
...
+
+  /**
+   * Aggregate func:compute the stddev of the specific column
+   *
+   * @group agg_funcs
+   */
+  def stddev(e: Column): Column = StdDeviation(e.expr)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: can you move these two functions to the aggregate function section, and change the comment to say

```
Aggregate function: Computes the standard deviation of the specific column.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala [hunk 4]
```diff
@@ -616,6 +617,115 @@ case class SumFunction(expr: Expression, base: AggregateExpression) extends Aggr
...
+      case DecimalType.Unlimited =>
+        DecimalType.Unlimited
+      case _ =>
+        DoubleType
+    }
+
+  private val zero = Cast(Literal(0), calcType)
+
+  private var count: Long = _
+
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala]: remove the extra blank line here

**Comment by rxin** [comment]
@dreamquster thanks for doing this -- I think we also need to add the Python version of it here as well.

@yhuai can you review the stddev expression code itself?

**Comment by yhuai** [comment]
@dreamquster Can you add comments in your code to explain:

1. Is it sample standard deviation or population standard deviation?
2. What is the algorithm used to calculate the standard deviation (how to calculate the partial result and how to calculate the final results)?

**Comment by dreamquster** [comment]
@yhuai . Is this comment OK?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala [hunk 5]
```diff
@@ -616,6 +617,119 @@ case class SumFunction(expr: Expression, base: AggregateExpression) extends Aggr
...
+    (seqPartialData, stddev)
+  }
+
+  override def toString: String = s"STDDEV($child)"
+
+  override def newInstance(): StdDeviationFunction = new StdDeviationFunction(child, this)
+}
+
+/**
+ * the standard deviation = sqrt(sum((xi-avg)^2)/N)
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala]: `N` -> `N - 1` for sample standard deviation. please check the result from R and make sure we output the same. And we should leave a TODO here to implement the numerically stable method in the future. See:

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala#L36

**Comment by marmbrus** [comment]
Thanks for working on this, but I think we should probably close this issue in favor of the implementation in #6297.

**Pull request closed without merging**